jax/_src/lax/search.py

```diff
@@ -0,0 +1,180 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from functools import partial
+import operator
+
+import numpy as np
+
+import jax
+from jax._src import ad_util
+from jax._src import api
+from jax._src import util
+from jax._src.lax import lax
+from jax._src.lax.control_flow import fori_loop
+from jax import core
+from jax.interpreters import ad
+from jax.interpreters import batching
+from jax.interpreters import xla
+
+
+def searchsorted(sorted_arr, query, *, side='left', dimension=0, batch_dims=0, method="default"):
+  """Find indices of query values within a sorted array.
+
+  Args:
+    sorted_arr : N-dimensional array, which is assumed to be sorted in increasing
+      order along ``dimension``.
+    query : N-dimensional array of query values.
+    side : {'left', 'right'}. If 'left', find the index of the first suitable
+      location. If 'right', find the index of the last.
+    dimension : integer specifying the dimension along which to insert query values.
+    batch_dims : integer specifying the number of leading dimensions of `sorted_arr`
+      and `query` to treat as batch dimensions.
+    method : {'default', 'scan', 'sort'} The method used to compute the result.
+      Outside JIT, this defaults to "scan". Within JIT, this defaults to "scan"
+      on CPU, and "sort" otherwise. Assume ``M = query.size`` and
+      ``N = sorted_arr.shape[dimension]``. ``method='scan'`` uses an *O[M log N]*
+      binary search algorithm, while ``method='sort'`` uses an *O[(M+N) log(M+N)]*
+      co-sorting approach. Although ``method='scan'`` may appear to have better
+      scaling overall, in practice it cannot take advantage of parallelism inherent
+      to accelerators, and so in most cases is slower than ``method='sort'`` on GPU
+      and TPU backends.
```

Comment:
At what point does it make sense to switch to parallel search (`(a < v).sum()`, vmapped over `v` is necessary) rather than bothering with a sort?

I don't know if this needs to be built in to searchsorted, but I'm sure this is faster on accelerators, at least for low dimensional inputs.

Reply:
Good question - I think that's hard to answer in general (different hardware will have different behavior) which is why this is difficult to do well without pushing the logic into the compiler.

Reply:
I gave this a try. To my surprise, in my tests using this for many vmapped interpolations on TPU, it didn't seem to really help.